rfcs/0049-flakes.md

```diff
@@ -0,0 +1,381 @@
+
+    outputs = { self, nixpkgs, grcov }: {
+      packages.grcov = stdenv.mkDerivation {
+        src = grcov;
+        ...
+      };
+    };
+
+The following input types are specified at present:
+
+* `git`: A Git repository or dirty local working tree.
+
+* `github`: A more efficient scheme to fetch repositories from GitHub
+  as tarballs. These have slightly different semantics from `git`
+  (in particular, the `revCount` attribute is not available).
+
+* `tarball`: A `.tar.{gz,xz,bz2}` file.
+
+* `path`: A directory in the file system. This generally should be
+  avoided in favor of `git` inputs, since `path` inputs have no
+  concept of revisions (only a content hash) or tracked files
+  (anything in the source directory is copied).
+
+* `hg`: A Mercurial repository.
+
+## Lock files
+
+Inputs specified in `flake.nix` are typically "unlocked" in that they
+don't specify an exact revision. To ensure reproducibility, Nix will
+automatically generate and use a *lock file* called `flake.lock` in
+the flake's directory. The lock file contains a tree of mappings from
+the inputs specified in `flake.nix` to inputs specifications that
+contain revisions.
+
+For example, if `flake.nix` has the inputs in the example above, then
+the resulting lock file might be:
+```
+{
+    "version": 4,
+    "inputs": {
+        "import-cargo": {
+            "inputs": {},
+            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",
+            "original": {
+              "type": "github",
+              "owner": "edolstra",
+              "repo": import-cargo"
+            },
+            "resolved": {
+              "type": "github",
+              "owner": "edolstra",
+              "repo": "import-cargo",
+              "rev": "c33e13881386931038d46a7aca4c9561144d582e"
+            }
+        },
+        "nixpkgs": {
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "original": {
+              "type": "github",
+              "owner": "NixOS",
+              "repo": nixpkgs"
+            },
+            "resolved": {
+              "type": "github",
+              "owner": "NixOS",
+              "repo": "nixpkgs",
+              "rev": "4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
+            }
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+revision `c33e13881386931038d46a7aca4c9561144d582e` of the
+`edolstra/import-cargo` repository on GitHub. Nix will also check that
+the content hash of the input is equal to the one recorded in the lock
+file. This check is superfluous for Git repositories (since the commit
+hash serves a similar purpose), but for GitHub archives, we cannot
+directly check that the contents match the commit hash.
+
+Note that lock files are only used at top-level: the `flake.lock`
+files in dependencies (if they exist) are ignored. The lock file
+transitively locks direct as well as indirect dependencies.
```

Comment:
It would be useful to have an option when fetching a flake as input to use the built version of that flake according to its own lockfile. But that can be a future enhancement.